### DISCUSSION TITLE
How to fix audio playback glitches?

### DISCUSSION THREAD

**Discussion by jubejubess** [jspsych/jsPsych#1066] [Q&A] [answered]
Hello!

I have an experiment that presents ~400 brief audio clips that are separated into 8 blocks. Each trial presents a small picture, an audio clip, then the keyboard response page with a visual prompt (click A OR B). Initially, we preloaded all the audio files at the beginning of the experiment, however the experiment didn't load/crashed for some people due to the large size of audio files that needed to be loaded. Now, I have it set up where it preloads the audio clip before each trial. The experiment loads properly, but occasionally (~10-20 out of ~400 trials) the audio playback lags from a few seconds to a minute or doesn't play at all, which overlaps or skips over the response page completely or the response page doesn't show up after the audio plays through. I was wondering if there was something that could help minimize the amount of audio or response page errors even further?

This is how the audio is preloaded for each trial:
````
var blocksize;
		blocksize = 50;

var preload = {
						type: "call-function",
						async: true,
						func: function (done) {
								jsPsych.pluginAPI.preloadAudioFiles([jsPsych.timelineVariable('stimulus', true)], function () {
										done({ preload: "success" });
								})
						}
				};

// there are 8 main procedures/blocks (400 total trials, separated into 8 blocks: 50 trials per block)
// this is just an example of how one block is preloaded

				var main_procedure_1 = {
					timeline: [
						preload, // preloading for first block (preloads before each trial?)
						fixation_n, // icon presentation
						audiopresentation, // audio presentation
						responselog // response page
					],
					timeline_variables: test_stimuli_1,
					randomize_order: true,
					sample : {
						type: 'without-replacement',
						size: blocksize
					}
				};

// the formatting of the fixation, audio and response variables

var fixation_1 = {
		type: 'image-keyboard-response',
		data: {test_part: 'fixation', blocktype: '1'},
		stimulus: 'shared_assets/img/1-small.png',
		prompt: '',
		choices: jsPsych.NO_KEYS,
		trial_duration: 500
	};

	var audiopresentation = {
		type: 'audio-keyboard-response',
		data: {test_part: 'test'},
		stimulus: jsPsych.timelineVariable('stimulus'),
		prompt: "",
		choices: jsPsych.NO_KEYS,
		trial_duration: 1100
	};

	var responselog = {
		type: 'html-keyboard-response',
		stimulus: "",
		data: jsPsych.timelineVariable('data'),
		prompt: function(){return "Do you hear <p>"+jsPsych.timelineVariable('segments',true);},
		choices: jsPsych.timelineVariable('choices'),
		on_finish: function(data){
			if(data.key_press){
				data.response = jsPsych.pluginAPI.convertKeyCodeToKeyCharacter(data.key_press);
			}/*
			else{
   				return
   			}*/
		},
		trial_duration: 15000
	};

````
 

If anyone has any insight, it would be much appreciated, thank you! :)

**Comment by becky-gilbert** [comment, marked as answer]
Hi @jubejubess, you might be running into the cache limits of your participant's browsers. This depends on the file sizes, and it's hard to know for sure because cache limits will depend on the browser and the amount of free memory on the computer. Your approach of loading the audio file before each trial makes sense. 

Have you checked that the preloading is working by opening your browser's network tab (e.g. [like this in Chrome](https://developers.google.com/web/tools/chrome-devtools/network))? You should see that each audio file is loaded only once, before the fixation and audio presentation part of the trial. You could increase the fixation duration if it's hard to tell when the file is loading. 

One potential issue with loading each file at the start of the trial is that this involves a lot of back-and-forth between the participant's computer and server, so it's vulnerable to intermittent connectivity problems. So another approach might be to move your call function so that it runs at the start of each block, and preload all 50 files that you need for the next block. This way you're loading fewer files than you were before, so hopefully you won't hit the browser's cache limit (if that was the original problem).

**Reply by jubejubess** [reply]
Hi Becky,

I checked and the files seem to be preloading properly. I'll try the approach to move the call-function to run at the start of each block and see if that works!

I just had another quick question. I tried converting the files from .wav to .mp3 to decrease the load that needs to be preloaded and I tested to see if I could preload them all at the beginning again. When I tested it with the network set to "Slow 3G" and occasionally "Fast 3G", I see that I receive that the experiment fails to load and in the console is the following error:

Uncaught TypeError: Cannot read property 'style' of null
    at update_loading_progress_bar (jspsych.js:2372)
    at Audio.<anonymous> (jspsych.js:2207)
However, the "experiment fails to load" screen and error appear before all the files have finished preloading and I see that the files continue to preload even after receiving this error. I was wondering if there was a particular reason why the files continue to preload even after it fails to load?

Thanks again for your help! :)

**Reply by becky-gilbert** [reply]
Great - yes do let us know if you end up solving this problem!

So I _think_ the 'Uncaught TypeError: Cannot read property 'style' of null' error that you're getting is a known bug (see #648). For instance, I see this error when I run the jspsych-audio-keyboard-response.html example file. It doesn't cause any performance issues and we'll fix it in the next release. 

However I'm not sure why the audio preloading would continue after you see the "Experiment failed to load" message, because that message should be shown when the preloading doesn't finish before the `max_load_time` duration that's set in `jsPsych.init`. I can look into this and get back to you.

**Reply by jubejubess** [reply]
@becky-gilbert Forgot to update about this, but we implemented preloading before each block instead and it's been working perfectly! Files were preloading as planned -- the wait time for preloading varied depending on their network speed, but no obvious crashing or glitching had occurred. Thanks again!

**Reply by becky-gilbert** [reply]
Great! And thanks very much for the update 😃
